features/430-passwordless/design.md

```diff
@@ -169,19 +183,97 @@ For examples:
+        3. Select and authenticate the user using the private key in key
+           cookie.
+        4. Authenticate as the user.
+- Native platform:
+    - App should store the key-pair securely.
+    - Signing up:
+        1. App generates key-pair, and put it in platform secure key store.
+        2. App request challenge from server.
+        2. App generates Attestion object using the challenge and key-pair, and
+           include it in authorization request as `login_hint`.
+        3. Server decode and verify `login_hint`, and register new anonymous
+           user using the public key.
+    - Logging in:
+        1. App retrieve key-pair from platform secure key store.
+        2. App request challenge from server.
+        2. App generates Assertion object using the challenge and key-pair, and
+           include it in authorization request as `login_hint`.
+        3. Server decode and verify `login_hint`, and authenticate existing
+           anonymous user using the public key.
+
+Two entities are introduced in the identity model:
+- Anonymous identity: consist of a key ID and public key
+- Anonymous authenticator: consist of a key ID and public key
+    - This is not designed as generic 'key-pair authenticator', since it has
+      different security properties from normal key-pair authentication: the
+      generation of key-pair is done on server-side, and private key may be
+      stored insecurely.
+
+For native platforms:
+- SDK should have ability to generate key-pairs
+- To enable browser-less anonymous authentication, a simple fetch with
+  follow-redirect should be able to complete authentication flow. SDK can
+  extract the authorization code from the response URL.
+
+## OIDC
+
+### `amr` claim
 
 To indicate authenticator used in authentication, `amr` claim is used in OIDC
 ID token.
 
 `amr` claim is an array of string. It includes authentication method used:
+- If secondary authentication is performed: `mfa` is included.
 - If password authenticator is used: `pwd` is included.
 - If any OTP (TOTP/OOB-OTP) is used: `otp` is included.
 - If WebAuthn is used: `hwk` is included.
 
 If no authentication method is to be included in `amr` claim, `amr` claim would
 be omitted from the ID token.
 
+### `acr` claim
+
+If any secondary authenticator is performed, `acr` claim would be included in
+ID token with value `http://schemas.openid.net/pape/policies/2007/06/multi-factor`.
+
+To perform step-up authentication, developer can pass a `acr_values` of 
+`http://schemas.openid.net/pape/policies/2007/06/multi-factor` to the
+authorize endpoint.
+
+### `login_hint` parameter
+
+Developer can optionally pre-select the identity to use using `login_hint`
+parameter. For examples:
+- To authenticate with email, `login_hint` should be
+  `login_id:email` or `login_id:email:<email address>`.
+- To authenticate with Google OAuth, `login_hint` should be `oauth:google` or
+  `oauth:google:<subject ID>`.
```

Comment:
Instead of `login_id:` and `oauth:` we should have `https://auth.skygear.io/login_hint?type=claims&email=email` and `https://auth.skygear.io/login_hint?type=attestation&data=...`. Using URL allow us to add extra fields in the future. Can we specify the exact format we are going to support?